auth/providers/azure/azure.go

```diff
@@ -79,19 +80,35 @@ type authInfo struct {
 	Issuer      string
 }
 
+var (
+	cachedAuthInfo *authInfo
+	mutex          = &sync.Mutex{}
+)
+
+// New is called per authentication request
 func New(ctx context.Context, opts Options) (auth.Interface, error) {
 	c := &Authenticator{
 		Options: opts,
 	}
-	authInfoVal, err := getAuthInfo(ctx, c.Environment, c.TenantID, c.HttpClientRetryCount, getMetadata)
-	if err != nil {
-		return nil, err
+
+	if cachedAuthInfo == nil {
+		mutex.Lock()
+		defer mutex.Unlock()
+		if cachedAuthInfo == nil {
```

Comment:
whats the reason for checking cachedAuthInfo == nil twice?

Reply:
making sure it's really nil once it gets the mutex because potentially it's already set by other thread once this mutex is acquired.